feat(dot/sync): improve worker pool

haikoschol opened the pull request:
The main difference in the worker pool API is that `SubmitBatch()` does not block until the whole batch has been processed. Instead, it returns an ID which can be used to retrieve the current state of the batch. In addition, `Results()` returns a channel over which task results are sent as they become available.

The main improvement this brings is increased concurrency, since results can be processed before the whole batch has been completed.

What has not changed is the overall flow of the Strategy interface; getting a new batch of tasks with `NextActions()` and processing the results with `Process()`.

## Changes

* replaced the code in `dot/sync/worker_pool.go`
* adapted `SyncService` to the API changes of the new worker pool
* adapted some expectations in tests regarding how often some mocks are called (hopefully without changing the logic being tested)

## Tests

```sh
go test github.com/ChainSafe/gossamer/dot/sync
```

## Issues

Closes #4198

haikoschol committed:
feat(dot/sync): improve worker pool

The main difference in the worker pool API is that SubmitBatch() does
not block until the whole batch has been processed. Instead, it returns
an ID which can be used to retrieve the current state of the batch.
In addition, Results() returns a channel over which task results are
sent as they become available.

The main improvement this brings is increased concurrency, since results
can be processed before the whole batch has been completed.

What has not changed is the overall flow of the Strategy interface;
getting a new batch of tasks with NextActions() and processing the
results with Process().

Closes #4232

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ChainSafe/gossamer?pullRequest=4258) <br/>All committers have signed the CLA.

haikoschol commented:
Created as a draft for two reasons:
1. I'd like to run a sync from scratch on Westend and/or Paseo as a regression test for a while.
2. To discuss and possibly address this TODO:
```
// SubmitBatch accepts a list of tasks and immediately returns a batch ID. The batch ID can be used to query the status
// of the batch using [GetBatchStatus].
// TODO
// If tasks are submitted faster than they are completed, resChan will run full, blocking the calling goroutine.
// Ideally this method would provide backpressure to the caller in that case. The rejected tasks should then stay in
// FullSyncStrategy.requestQueue until the next round. But this would need to be supported in all sync strategies.
func (w *workerPool) SubmitBatch(tasks []Task) (id BatchID, err error) {
```

dimartiro commented on dot/sync/service.go:
```diff
@@ -119,6 +135,11 @@ func NewSyncService(cfgs ...ServiceConfig) *SyncService {
+		workerPool: NewWorkerPool(WorkerPoolConfig{
+			MaxRetries: 5,
+			// TODO: This should depend on the actual configuration of the currently used sync strategy.
+			Capacity: defaultNumOfTasks * 10,
```
Why times 10?

haikoschol replied:
Just an arbitrary value that "should be enough" while I was testing the branch on the Westend genesis instance. Apart from wanting to do this testing, this TODO and the other one about back pressure are the reasons why this is still a draft.
Do you think it makes sense to use the actual value for number of tasks configured in the sync strategy? And should we add some extra space here or not?

dimartiro replied:
Ok got it, we can move it to a constant to self explain the usage.

haikoschol replied:
I'd like to try and make this depend on the configuration of the strategy. But if that's too invasive, I'll move it to a constant.

haikoschol replied:
I've done this in [b6cf4d7](https://github.com/ChainSafe/gossamer/pull/4258/commits/b6cf4d781f5764f92c4a9be1298e2a832d999f2d). It required extending the `Strategy` interface and delaying initialization of the worker pool.

dimartiro commented on dot/sync/service.go:
```diff
@@ -88,12 +89,27 @@ type Change struct {
 	IsSynced() bool
 }
 
+type syncTask struct {
```
Could you please move this to `fullsync.go` since it is specific for that strategy?

dimartiro commented on dot/sync/service.go:
```diff
@@ -88,12 +89,27 @@ type Change struct {
+}
+
+func (s *syncTask) ID() TaskID {
+	return TaskID(s.request.String())
```
I'm not sure if this is a good ID since it is the string representation of the request, if we send the same request for multiple peers we are gonna get the same ID here.
What if we just generate an UUID?
Also, we can add a `String()` for the string representation that could be useful for debugging purposes

haikoschol replied:
Good point! I had thought about whether a timestamp is a good ID for the batch, but hadn't considered this one. Just found the `String()` method and ran with it.
A UUID is a bit unwieldy, but we already use a library for generating them and I don't have a better idea, so went with that.
Also added `String()` and used it in logging calls.

dimartiro commented on dot/sync/service.go:
```diff
@@ -119,6 +135,11 @@ func NewSyncService(cfgs ...ServiceConfig) *SyncService {
 		stopCh:                make(chan struct{}),
 		seenBlockSyncRequests: lrucache.NewLRUCache[common.Hash, uint](100),
+		workerPool: NewWorkerPool(WorkerPoolConfig{
+			MaxRetries: 5,
```
Could you move the magic numbers to a const? 😄

dimartiro commented on dot/sync/fullsync.go:
```diff
@@ -395,85 +407,79 @@ type RequestResponseData struct {
-					},
-				})
-			}
+	task, ok := result.Task.(*syncTask)
```
What if we define `result.Task` over a generic? so we can skip this casting?

haikoschol replied:
I considered making this generic and decided against it, but I can't remember why. Will give it a try.

haikoschol replied:
Just tried and the reason I didn't do this is because `SyncService` is supposed to maintain the worker pool and work with any kind of `Strategy`. This won't work if the worker pool is a generic type instantiated with `*syncTask`.

With the current implementation, only the strategies themselves know the type of the tasks processed by the worker pool.

dimartiro replied:
Good point, that makes me think that the strategies should be responsible for creating their own workerpool maybe

haikoschol replied:
I'm still in favor of having the service manage one pool that is used for all strategies. I think it's easier to manage in terms of concurrency and good separation of concern. Is it worth giving that up to avoid a few casts? The strategies definitely know what concrete type they use as sync tasks.

dimartiro replied:
100% agree that the service should be responsible for the working pool, but I'm still trying to find a way to improve the types. We can keep the casts in the meanwhile

dimartiro commented on dot/sync/fullsync.go:
```diff
@@ -395,85 +407,79 @@ type RequestResponseData struct {
 
-			continue
-		}
+	request := task.request.(*messages.BlockRequestMessage)
```
you can skip this cast if you change the `syncTask.request` from `messages.P2PMessage` to `*messages.BlockRequestMessage` since (if i'm not wrong) this is the only type we are expecting

haikoschol replied:
Yes, if we consider `syncTask` specific to `FullSyncStrategy`, we can narrow down the type here.

haikoschol committed:
chore(dot/sync): move syncTask to fullsync.go

haikoschol committed:
feat(dot/sync): use UUID for syncTask.ID()

haikoschol committed:
chore(dot/sync): use constant for maxTaskRetries

haikoschol committed:
chore(dot/sync): narrow the type of syncTask.request

haikoschol committed:
feat(dot/sync): make worker pool capacity depend on strategy config

haikoschol committed:
feat(dot/sync): avoid overwhelming the worker pool

When the worker pool falls behind processing tasks, the service won't
ask the strategy for more tasks and instead directly runs Process()
again.

haikoschol commented:
> 2. To discuss and possibly address this TODO:

I've addressed the TODO in [f4a9ccc](https://github.com/ChainSafe/gossamer/pull/4258/commits/f4a9cccf47b5a9fbde6ef73bd82ea631b1fea27c). Not sure if this is the right approach, but I think it is better than always adding new tasks, risking a blocked goroutine.

haikoschol committed:
remove unused error return value from WorkerPool.SubmitBatch()

haikoschol committed:
don't limit retries on sync tasks

It doesn't make sense to give up since we need the blocks in that
request to progress.

haikoschol committed:
remove peers with network errors from worker pool

haikoschol committed:
ensure ID is included in syncTask.String()

haikoschol committed:
WIP

haikoschol commented:
We [decided](https://github.com/ChainSafe/gossamer/issues/4198#issuecomment-2499560893) not to proceed with this.
